src/communication/CommandLine.cpp

```diff
@@ -0,0 +1,475 @@
+    Scene& scene = Application::getInstance().current_slice->scene;
+    if (document.HasMember("metadata") && document["metadata"].IsObject())
+    {
+        const rapidjson::Value& metadata = document["metadata"];
+        if (metadata.HasMember("machine_extruder_trains") && metadata["machine_extruder_trains"].IsObject())
+        {
+            const rapidjson::Value& extruder_trains = metadata["machine_extruder_trains"];
+            for (rapidjson::Value::ConstMemberIterator extruder_train = extruder_trains.MemberBegin(); extruder_train != extruder_trains.MemberEnd(); extruder_train++)
+            {
+                const int extruder_nr = atoi(extruder_train->name.GetString());
+                if (extruder_nr < 0)
+                {
+                    continue;
+                }
+                while (scene.extruders.size() <= static_cast<size_t>(extruder_nr))
+                {
+                    scene.extruders.emplace_back(scene.extruders.size(), &scene.settings);
+                }
+                const rapidjson::Value& extruder_id = extruder_train->value;
+                if (!extruder_id.IsString())
+                {
+                    continue;
+                }
+                const std::string extruder_definition_id(extruder_id.GetString());
+                const std::string extruder_file = findDefinitionFile(extruder_definition_id, search_directories);
+                loadJSON(extruder_file, scene.extruders[extruder_nr].settings);
+            }
+        }
+    }
+
+    if (document.HasMember("settings") && document["settings"].IsObject())
+    {
+        loadJSONSettings(document["settings"], settings);
+    }
+    if (document.HasMember("overrides") && document["overrides"].IsObject())
+    {
+        loadJSONSettings(document["overrides"], settings);
+    }
+    return 0;
+}
+
+void CommandLine::loadJSONSettings(const rapidjson::Value& element, Settings& settings)
+{
+    for (rapidjson::Value::ConstMemberIterator setting = element.MemberBegin(); setting != element.MemberEnd(); setting++)
+    {
+        const std::string name = setting->name.GetString();
+
+        const rapidjson::Value& setting_object = setting->value;
+        if (!setting_object.IsObject())
+        {
+            logError("JSON setting %s is not an object!\n", name.c_str());
+            continue;
+        }
+
+        if (setting_object.HasMember("children"))
+        {
+            loadJSONSettings(setting_object["children"], settings);
+        }
+        else //Only process leaf settings. We don't process categories or settings that have sub-settings.
+        {
+            if (!setting_object.HasMember("default_value"))
+            {
+                logWarning("JSON setting %s has no default_value!\n", name.c_str());
+                continue;
```

Comment:
The way I remember it all settings are required to have a default value.

You might want to `std::exit(1)` here so that we can find the problem early and easily.

Reply:
It was breaking on a setting that is used internally by the front-end only. I gave that setting a default now (https://github.com/Ultimaker/Cura/commit/1fbf9c973145c202ecc6773d18e70234ca6af8d4) but since you'll see a warning I don't think it would be useful to break on that sort of thing. If the setting is actually used, it'll exit when it's being used.

Reply:
But that setting is not in the json files, right?

I was thinking that anything in the json files should always have a default_value. However, passing an unknown setting to the engine should result in a warning rather than an error. That makes it a lot easier to make a new feature for CuraEngine while you haven't made the setting yet.

Reply:
That setting was in the JSON files. In fdmprinter.def.json in any case.

We don't give warnings for settings that are not used by the engine. There are many such settings, such as `machine_buildplate_type`, `machine_steps_per_mm_y` or `material_surface_energy`.